### PR TITLE
feat: automatically do multi-part upload on 90+ MB files

### PR DESCRIPTION
**Pull request by chamini2**
image

### CODE REVIEW

File: libs/client/package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
   "types": "./src/index.d.ts",
   "dependencies": {
     "@msgpack/msgpack": "^3.0.0-beta2",
+    "async-sema": "^3.1.1",
```
Comment (chamini2) [inline libs/client/package.json:33]: i am not sure where I should add this @drochetti

File: libs/client/src/storage.ts [hunk 2]
```diff
@@ -76,6 +74,161 @@ async function initiateUpload(
...
+  let responses: MultipartObject[];
+  try {
+    // Does this wait for all to finish even if an early one fails?
+    responses = await Promise.all(partPromises);
+  } catch (error) {
+    // If any of the parts fail, cancel other uploads
+    cancelled.current = true;
+
+    console.error("Multipart upload failed, aborting upload", error);
+    throw error;
```
Comment (drochetti) [inline libs/client/src/storage.ts:161]: I think it's safe to remove console.error and either re-throw the error or throw a new one with the custom message + the cause

File: libs/client/src/storage.ts [hunk 3]
```diff
@@ -76,6 +74,161 @@ async function initiateUpload(
...
+  }
+
+  console.log("All parts uploaded, completing upload", responses);
```
Comment (drochetti) [inline libs/client/src/storage.ts:164]: remove?

File: libs/client/src/storage.ts [hunk 4]
```diff
@@ -76,6 +74,161 @@ async function initiateUpload(
...
+
+  try {
+    const response = await fetch(uploadUrl, {
+      method: "PUT",
+      body: chunk,
+    });
+
+    return (await responseHandler(response)) as MultipartObject;
+  } catch (error) {
+    console.error("Part upload failed, retrying", uploadUrl, error);
```
Comment (drochetti) [inline libs/client/src/storage.ts:124]: Safe to remove?

File: libs/client/src/storage.ts [hunk 5]
```diff
@@ -76,6 +74,161 @@ async function initiateUpload(
...
+  partNumber: number;
+  etag: string;
+};
+
+async function partUploadRetries(
+  uploadUrl: string,
+  chunk: Blob,
+  config: RequiredConfig,
+  cancelled: { current: boolean },
+  tries: number,
```
Comment (drochetti) [inline libs/client/src/storage.ts]: nit: `attemptsLeft`, with a default value

File: libs/client/src/storage.ts [hunk 6]
```diff
@@ -76,6 +74,161 @@ async function initiateUpload(
...
+  const response = await fetch(completeUrl, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      parts: responses.map((mpart) => ({
+        partNumber: mpart.partNumber,
+        etag: mpart.etag,
+      })),
```
Comment (drochetti) [inline libs/client/src/storage.ts:176]: Isn't `responses` already what `parts` expect?

**Review by drochetti — COMMENTED** [review]
Thanks for handling this, unlocking large file sizeeees 🚀 

I left a few minor comments, other than the `async-sema` we discussed offline. Everything else lgtm!

**Review by drochetti — APPROVED** [review]

**Pull request merged**